rasa/shared/nlu/training_data/features.py

```diff
@@ -115,3 +118,16 @@ def __eq__(self, other: Any) -> bool:
             and other.attribute == self.attribute
             and other.features == self.features
         )
+
+    def fingerprint(self) -> Text:
+        """Calculate a stable string fingerprint for the features."""
+        if isinstance(self.features, np.ndarray):
+            f_as_text = self.features.tostring()
+        else:
```

Comment:
How about doing that during instance initialization?

Reply:
or both? :P

Reply:
Hm, thinking about this again - it seems somewhat against Python's duck typing paradigm doesn't it? I am against a warning, as we shouldn't just warn about something that is actually an error. And if we raise, we have more control about the error message, but actually`AttributeError: 'list' object has no attribute 'tostring'` or `AttributeError: 'list' object has no attribute 'tocoo'` would be good enough wouldn't they?

Reply:
I just think it's not a very helpful error message. If we know we only support `ndarray` and `spmatrix` and we are already doing a type check, why not raise a more informative error if `self.features` is something else? (Also worth remembering that the community can write their own featurisers)

Reply:
I think the option that is most flexible and gives the users most opportunities - although it does not exactly adhere to our type annotations would be the following (which would allow for a list of list instead of an np.array for example): 

```
        if self.is_dense():
            f_as_text = self.features.tostring()
        elif self.is_sparse():
            f_as_text = rasa.shared.nlu.training_data.util.sparse_matrix_to_string(
                self.features
            )
        else:
            f_as_text = str(self.features)
```

While we could enforce the types, I don't think the `fingerprint()` method should be the place where we enforce the types. The type checks there are just to ensure sensible behavior. Also, I don't think enforcing types is a common thing in our codebase as it runs against the duck typing paradigm

Another option for users is to extend the features type - although I am not sure that would be an viable option for downstream use

Reply:
the `is_dense` and `is_sparse` methods actually also depend on the right types to work as expected. Idk, I think if we want to enforce types we should do it right at the initialization to fail early and not wait for some method invocation.

Reply:
I actually hadn't realised that we already had `features: Union[np.ndarray, scipy.sparse.spmatrix]` in the `__init__`.
Then I think it's fine as is 👍